cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java

```diff
@@ -159,6 +165,42 @@ public ProgramStatus getProgramStatus(ProgramId programId) throws Exception {
     return getExistingAppProgramStatus(appSpec, programId);
   }
 
+  /**
+   * Returns the program run count of the given program.
+   *
+   * @param programId the id of the program for which the count call is made
+   * @return the run count of the program
+   * @throws NotFoundException if the application to which this program belongs was not found or the program is not
+   *                           found in the app
+   */
+  public int getProgramRunCount(ProgramId programId) throws Exception {
+    AuthorizationUtil.ensureAccess(programId, authorizationEnforcer, authenticationContext.getPrincipal());
+    return store.getProgramRunCount(programId);
+  }
+
+  /**
+   * Returns the program run count of the given program id list.
+   *
+   * @param programIds the list of program ids to get the count
+   * @return the counts of given program ids
+   */
+  public List<RunCountResult> getProgramRunCounts(List<ProgramId> programIds) throws Exception {
+    List<RunCountResult> result = store.getProgramRunCounts(programIds);
+
+    // filter the result
+    Set<? extends EntityId> visibleEntities = authorizationEnforcer.isVisible(new HashSet<>(programIds),
+                                                                              authenticationContext.getPrincipal());
+    for (int i = 0; i < result.size(); i++) {
+      RunCountResult runCountResult = result.get(i);
+      ProgramId programId = runCountResult.getProgramId();
+      if (!visibleEntities.contains(programId)) {
+        result.set(i, new RunCountResult(programId, null,
```

Comment:
Shouldn't modify the list, especially if the list was not "new" locally. Do a map and collect instead.

```java
return result.stream()
  .map(runCount -> {
    if (!visibleEntities.contains(runCount.getProgramId())) {
      return new RunCountResult(runCount.getProgramId(), null, new UnauthorizedException(...));
    }
    return runCount;
  })
  .collect(Collectors.toList());
```